[material-ui][docs] Add container queries docs (#42434)
mui/material-ui: docs/data/material/customization/containerQueries/containerQueries.md

# Container Queries

> <p class="description">API that enables the use of CSS container queries based on theme breakpoints.</p>

Reviewer: ```suggestion
<p class="description">A utility function based on the theme breakpoints for creating CSS container queries.</p>
```

Effectively doing this!

> A utility function based on the [theme breakpoints](/material-ui/customization/breakpoints/) for creating [CSS container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries).

Reviewer: I'd put this sentence as the page's description and then use the first sentence of the "Usage" section to add the MDN link (given the link to theme.breakpoints is already there. And then, remove this line here.

Author: Got it.

## Usage

To create a container query, use `theme.containerQueries` with any method available in the [theme.breakpoints](/material-ui/customization/breakpoints/#api).

> The value can be a unitless (pixel), a string, or a breakpoint key, for example:

Reviewer: ```suggestion
The value can be unitless (rendered in pixels), a string, or a breakpoint key. For example:
```

For clarity regarding the unitless value

```js
theme.containerQueries.up('sm'); // => '@container (min-width: 600px)'
```

:::info
One of the ancestors must have the CSS container type specified.
:::

{{"demo": "BasicContainerQueries.js"}}

### Named containment contexts

> To refer to a [containment context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries#naming_containment_contexts), call the `containerQueries` method with the name, then you will be able to access all the breakpoint methods like the example above:

Reviewer: ```suggestion
To refer to a [containment context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries#naming_containment_contexts), call the `containerQueries` method with the container name. 
That's how you can access all the breakpoint methods like the example above:
```

I'm assuming "the name" refers to the container name? And a tiny tweak here to remove the future tense.

```js
theme.containerQueries('sidebar').up('500px'); // => '@container sidebar (min-width: 500px)'
```

## Shorthand syntax

> In sx prop, a format of `@<size>` or `@<size>/<name>` can be used to apply container queries without refering to the theme.

Reviewer: ```suggestion
When adding styles using the `sx` prop, use the `@<size>` or `@<size>/<name>` notation to apply container queries without referring to the theme.
```

> In sx prop, a format of `@<size>` or `@<size>/<name>` can be used to apply container queries without refering to the theme.

Reviewer 2: Unrelated: I can't wait for a future where we can use the same syntax in `styled` and `sx`.

Author: It sounds possible to me for v7 (only with Pigment CSS).

- `<size>`: a width or a breakpoint key.
- `<name>` (optional): a named containment context.

For example:

```js
<Card
  sx={{
    display: 'flex',
    flexDirection: {
      '@': 'column', // @container (min-width: 0px)
      '@250': 'row', // @container (min-width: 250px)
    },
  }}
/>
```

{{"demo": "SxPropContainerQueries.js"}}

### Caveats

- When `@` is used without a width or name, `0px` is applied.

> - To get pixel unit, use a unitless number. For example, `@500` is equivalent to `500px`. Using `@500px` will not work.

Reviewer: ```suggestion
- To render as a pixel, use a unitless number. For example, `@500` is equivalent to `500px`. Using `@500px` does not work.
```

> - Container queries, **with the same unit**, are sorted in ascending order.

Author: My intention is to say "same unit, order does not matter".

So, I changed to "Container queries must share the same units (the sizes can be defined in any order), as shown below:"

Is it okay?
